{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.00 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 0.08 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.00 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.00 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.04 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index data/MG1655-K12.first400K.fasta\n",
      "[main] Real time: 0.141 sec; CPU: 0.125 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48301, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.27, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.673 CPU sec, 3.598 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48357, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.22, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.808 CPU sec, 3.650 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48345, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.442 CPU sec, 3.320 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48337, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.37, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.406 CPU sec, 3.287 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48338, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.408 CPU sec, 3.281 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48367, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.33, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.462 CPU sec, 3.316 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48287, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.23, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.965 CPU sec, 3.851 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48328, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.986 CPU sec, 3.742 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48359, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.37, 10.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.499 CPU sec, 3.345 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48323, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.38, 10.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.566 CPU sec, 3.413 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48276, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.26, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mem_process_seqs] Processed 100000 reads in 3.533 CPU sec, 3.393 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48231, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.27, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.524 CPU sec, 3.394 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48187, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.612 CPU sec, 3.507 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48265, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.41, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.542 CPU sec, 3.407 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48287, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.396 CPU sec, 3.275 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48307, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.33, 10.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.346 CPU sec, 3.214 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48303, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.40, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.775 CPU sec, 3.617 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48393, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.40, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 5.024 CPU sec, 4.799 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48320, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 5.027 CPU sec, 4.875 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48377, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.26, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.487 CPU sec, 3.340 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48307, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.392 CPU sec, 3.271 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48322, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.27, 10.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.205 CPU sec, 3.067 real sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48361, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.23, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.227 CPU sec, 3.095 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48350, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.320 CPU sec, 3.178 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48319, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.156 CPU sec, 3.024 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48334, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.161 CPU sec, 3.026 real sec\n",
      "[M::process] read 63204 sequences (6320400 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48289, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.28, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.226 CPU sec, 3.128 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 30551, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 63204 reads in 2.005 CPU sec, 1.915 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem data/MG1655-K12.first400K.fasta data/ecoli_400K_err_1.fastq data/ecoli_400K_err_2.fastq\n",
      "[main] Real time: 95.518 sec; CPU: 99.307 sec\n"
     ]
    }
   ],
   "source": [
    "!bwa index data/MG1655-K12.first400K.fasta\n",
    "!bwa mem data/MG1655-K12.first400K.fasta data/ecoli_400K_err_1.fastq data/ecoli_400K_err_2.fastq > ecoli_400.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cnt_err_del(sam, fasta, n_as_error=False, start_positions=None):\n",
    "\n",
    "    with open(fasta) as file:\n",
    "        reference = ''.join([line.strip() for line in file if not line.startswith('>')])\n",
    "    ref_length = len(reference)\n",
    "    \n",
    "    \n",
    "    n_errors, n_deletions = {}, {}\n",
    "    correct_positions = 0\n",
    "    \n",
    "    for num, line in enumerate(open(sam)):\n",
    "        if line.startswith('@'):\n",
    "            continue\n",
    "            \n",
    "        data = line.strip().split()\n",
    "        if data[9] == '*':\n",
    "            continue\n",
    "        start = int(data[3])\n",
    "        if not start:\n",
    "            continue\n",
    "            \n",
    "        read_name = data[0]\n",
    "        read_type = 'forward' if int(data[8]) > 0 else 'backward'\n",
    "        \n",
    "        if start_positions and (read_name, read_type) in start_positions:\n",
    "            start_ecoli_pos = start_positions[(read_name, read_type)]\n",
    "            if np.abs(start_ecoli_pos - start) > 100:\n",
    "                continue\n",
    "                \n",
    "            start = start_ecoli_pos\n",
    "            \n",
    "        if not start:\n",
    "            continue\n",
    "\n",
    "        errors = {}\n",
    "        deletions = {}\n",
    "        read = list(data[9])\n",
    "        read_length = len(read)\n",
    "        reference_region = reference[start - 1 : min(start - 1 + read_length, ref_length)]\n",
    "        \n",
    "        for pos, r in enumerate(reference_region):\n",
    "            if r != read[pos]:\n",
    "                if read[pos] == 'N' and not n_as_error:\n",
    "                    deletions[start + pos] = (r, read[pos])\n",
    "                else:\n",
    "                    errors[start + pos] = (r, read[pos])\n",
    "            else:\n",
    "                correct_positions += 1\n",
    "                \n",
    "        if len(errors) > 0:\n",
    "            n_errors[(read_name, read_type)] = errors\n",
    "            \n",
    "        if len(deletions) > 0:\n",
    "            n_deletions[(read_name, read_type)] = deletions\n",
    "    return n_errors, n_deletions, correct_positions\n",
    "\n",
    "\n",
    "def read_names(sam):\n",
    "    names = set()\n",
    "    for num, line in enumerate(open(sam)):\n",
    "        if line.startswith('@'):\n",
    "            continue\n",
    "            \n",
    "        data = line.strip().split()\n",
    "        if data[2] == '*' or data[5] == '*':\n",
    "            continue\n",
    "            \n",
    "        start = int(data[3])\n",
    "        if not start:\n",
    "            continue\n",
    "            \n",
    "        read_name = data[0]\n",
    "        read_type = 'forward' if int(data[8]) > 0 else 'backward'\n",
    "        names.add((read_name, read_type))\n",
    "    return names\n",
    "\n",
    "\n",
    "def deleted(total_reads, sam):\n",
    "    deleted_reads = total_reads.copy()\n",
    "    for num, line in enumerate(open(sam)):\n",
    "        if line.startswith('@'):\n",
    "            continue\n",
    "            \n",
    "        data = line.strip().split()\n",
    "        if data[2] == '*' or data[5] == '*':\n",
    "            continue\n",
    "            \n",
    "        start = int(data[3])\n",
    "        if not start:\n",
    "            continue\n",
    "            \n",
    "        read_name = data[0]\n",
    "        read_type = 'forward' if int(data[8]) > 0 else 'backward'\n",
    "        if (read_name, read_type) in total_reads:\n",
    "            deleted_reads.remove((read_name, read_type))\n",
    "    return deleted_reads\n",
    "\n",
    "\n",
    "def deleted_correct_positions(errors, deletions, deleted_reads, sam):\n",
    "    n = 0\n",
    "    for num, line in enumerate(open(sam)):\n",
    "        if line.startswith('@'):\n",
    "            continue\n",
    "            \n",
    "        data = line.strip().split()\n",
    "        if data[2] == '*' or data[5] == '*':\n",
    "            continue\n",
    "            \n",
    "        start = int(data[3])\n",
    "        if not start:\n",
    "            continue\n",
    "            \n",
    "        read_name = data[0]\n",
    "        read_type = 'forward' if int(data[8]) > 0 else 'backward'\n",
    "        read = list(data[9])\n",
    "        \n",
    "        if (read_name, read_type) in deleted_reads:\n",
    "            read_length = len(read)\n",
    "            if (read_name, read_type) not in errors:\n",
    "                n += read_length\n",
    "                continue\n",
    "                \n",
    "            errors_n = len(errors[(read_name, read_type)])\n",
    "            n += (read_length - errors_n)\n",
    "            \n",
    "        elif (read_name, read_type) in deletions:\n",
    "            errors_positions = errors[(read_name, read_type)] if (read_name, read_type) in errors else {}\n",
    "            for pos in deletions[(read_name, read_type)]:\n",
    "                if pos not in errors_positions:\n",
    "                    n += 1\n",
    "    return n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Trimmomatic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TrimmomaticPE: Started with arguments:\n",
      " data/ecoli_400K_err_1.fastq data/ecoli_400K_err_2.fastq trim1_paired.fq trim1_unpaired.fq trim2_paired.fq trim2_unpaired.fq LEADING:5 TRAILING:5 SLIDINGWINDOW:4:15 MINLEN:35\n",
      "Multiple cores found: Using 4 threads\n",
      "Quality encoding detected as phred33\n",
      "Input Read Pairs: 1381602 Both Surviving: 1314087 (95,11%) Forward Only Surviving: 36625 (2,65%) Reverse Only Surviving: 26152 (1,89%) Dropped: 4738 (0,34%)\n",
      "TrimmomaticPE: Completed successfully\n"
     ]
    }
   ],
   "source": [
    "!trimmomatic PE data/ecoli_400K_err_1.fastq data/ecoli_400K_err_2.fastq \\\n",
    "trim1_paired.fq trim1_unpaired.fq trim2_paired.fq trim2_unpaired.fq \\\n",
    "LEADING:5 TRAILING:5 SLIDINGWINDOW:4:15 MINLEN:35"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 109032 sequences (10000046 bp)...\n",
      "[M::process] read 109552 sequences (10000124 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 52822, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.25, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 109032 reads in 3.460 CPU sec, 3.388 real sec\n",
      "[M::process] read 109748 sequences (10000159 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53112, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.26, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 109552 reads in 3.167 CPU sec, 2.940 real sec\n",
      "[M::process] read 110184 sequences (10000136 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53200, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 109748 reads in 3.263 CPU sec, 3.129 real sec\n",
      "[M::process] read 110674 sequences (10000080 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53438, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 110184 reads in 2.859 CPU sec, 2.706 real sec\n",
      "[M::process] read 111610 sequences (10000107 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53676, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 110674 reads in 3.083 CPU sec, 2.951 real sec\n",
      "[M::process] read 110332 sequences (10000031 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54111, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.23, 10.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 111610 reads in 3.321 CPU sec, 3.176 real sec\n",
      "[M::process] read 111160 sequences (10000009 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53465, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 110332 reads in 3.032 CPU sec, 2.882 real sec\n",
      "[M::process] read 111872 sequences (10000069 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 53922, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.39, 10.09)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 111160 reads in 3.621 CPU sec, 3.480 real sec\n",
      "[M::process] read 111878 sequences (10000190 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54232, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 111872 reads in 3.015 CPU sec, 2.864 real sec\n",
      "[M::process] read 113266 sequences (10000036 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54211, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.23, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 111878 reads in 3.392 CPU sec, 3.252 real sec\n",
      "[M::process] read 112038 sequences (10000014 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54862, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 113266 reads in 3.056 CPU sec, 2.903 real sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::process] read 109106 sequences (10000148 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 54275, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 112038 reads in 3.216 CPU sec, 3.061 real sec\n",
      "[M::process] read 108334 sequences (10000075 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 52845, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.28, 10.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 109106 reads in 2.988 CPU sec, 2.840 real sec\n",
      "[M::process] read 107532 sequences (10000135 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 52527, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 108334 reads in 2.886 CPU sec, 2.743 real sec\n",
      "[M::process] read 106940 sequences (10000059 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 52128, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.42, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 107532 reads in 3.445 CPU sec, 3.302 real sec\n",
      "[M::process] read 106390 sequences (10000148 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51857, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 106940 reads in 2.918 CPU sec, 2.761 real sec\n",
      "[M::process] read 106512 sequences (10000045 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51567, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.25, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 106390 reads in 2.968 CPU sec, 2.828 real sec\n",
      "[M::process] read 106508 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51595, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.30, 10.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 106512 reads in 2.888 CPU sec, 2.736 real sec\n",
      "[M::process] read 106630 sequences (10000097 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51580, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.26, 10.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 106508 reads in 2.934 CPU sec, 2.772 real sec\n",
      "[M::process] read 106206 sequences (10000108 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51701, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 106630 reads in 2.878 CPU sec, 2.717 real sec\n",
      "[M::process] read 105886 sequences (10000122 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51460, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 106206 reads in 2.895 CPU sec, 2.737 real sec\n",
      "[M::process] read 105670 sequences (10000022 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51287, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.28, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 105886 reads in 2.906 CPU sec, 2.740 real sec\n",
      "[M::process] read 105478 sequences (10000022 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51150, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.26, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mem_process_seqs] Processed 105670 reads in 2.934 CPU sec, 2.767 real sec\n",
      "[M::process] read 15636 sequences (1488812 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 51099, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.26, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 105478 reads in 2.901 CPU sec, 2.790 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 7576, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.29, 10.30)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 15636 reads in 0.627 CPU sec, 0.499 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem data/MG1655-K12.first400K.fasta trim1_paired.fq trim2_paired.fq\n",
      "[main] Real time: 71.249 sec; CPU: 74.763 sec\n"
     ]
    }
   ],
   "source": [
    "!bwa mem data/MG1655-K12.first400K.fasta trim1_paired.fq trim2_paired.fq > trim_paired.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_errors, raw_deletions, raw_correct_positions = cnt_err_del('ecoli_400.sam', 'data/MG1655-K12.first400K.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "trim_errors, trim_deletions, trim_correct_positions = cnt_err_del('trim_paired.sam', \n",
    "                                                                  'data/MG1655-K12.first400K.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_raw_errors = 0\n",
    "for positions in raw_errors.values():\n",
    "    total_raw_errors += len(positions)\n",
    "total_trim_errors = 0\n",
    "for positions in trim_errors.values():\n",
    "    total_trim_errors += len(positions)\n",
    "    \n",
    "result = pd.DataFrame(columns=['Error in corrected reads', 'Correct base in corrected reads', \n",
    "                               'Base is absent in corrected reads'], \n",
    "                      index=['Error in raw data', 'Correct base in raw data'])\n",
    "\n",
    "result.loc['Error in raw data', 'Error in corrected reads'] = f'{total_trim_errors} ({total_trim_errors / total_raw_errors * 100:.3f}%)'\n",
    "\n",
    "result.loc['Error in raw data', 'Correct base in corrected reads'] = f'0'\n",
    "\n",
    "result.loc['Error in raw data', 'Base is absent in corrected reads'] = f'{total_raw_errors - total_trim_errors} ({(total_raw_errors - total_trim_errors) / total_raw_errors * 100:.3f}%)'\n",
    "\n",
    "result.loc['Correct base in raw data', 'Error in corrected reads'] = f'0'\n",
    "\n",
    "result.loc['Correct base in raw data', 'Correct base in corrected reads'] = f'{trim_correct_positions} ({trim_correct_positions / raw_correct_positions * 100:.3f}%)'\n",
    "\n",
    "result.loc['Correct base in raw data', 'Base is absent in corrected reads'] = f'{raw_correct_positions - trim_correct_positions} ({(raw_correct_positions - trim_correct_positions) / raw_correct_positions * 100:.3f}%)'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Error in corrected reads</th>\n",
       "      <th>Correct base in corrected reads</th>\n",
       "      <th>Base is absent in corrected reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Error in raw data</td>\n",
       "      <td>1389165 (9.354%)</td>\n",
       "      <td>0</td>\n",
       "      <td>13462255 (90.646%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Correct base in raw data</td>\n",
       "      <td>0</td>\n",
       "      <td>239971114 (91.883%)</td>\n",
       "      <td>21197833 (8.117%)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Error in corrected reads  \\\n",
       "Error in raw data                1389165 (9.354%)   \n",
       "Correct base in raw data                        0   \n",
       "\n",
       "                         Correct base in corrected reads  \\\n",
       "Error in raw data                                      0   \n",
       "Correct base in raw data             239971114 (91.883%)   \n",
       "\n",
       "                         Base is absent in corrected reads  \n",
       "Error in raw data                       13462255 (90.646%)  \n",
       "Correct base in raw data                 21197833 (8.117%)  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Spades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def spades_table(errors1, errors2, deletions2, sam1, sam2, good1, good2):\n",
    "\n",
    "    first_reads = read_names(sam1)\n",
    "    second_deleted_reads = deleted(first_reads, sam2)\n",
    "    \n",
    "    total_reads = len(first_reads)\n",
    "    deleted_reads = len(second_deleted_reads)\n",
    "    n_errors = 0\n",
    "    corrected_errors = 0\n",
    "    deleted_errors = 0\n",
    "    \n",
    "    for (read, read_type), positions in errors1.items():\n",
    "        n = len(positions)\n",
    "        n_errors += n\n",
    "        if (read, read_type) in second_deleted_reads:\n",
    "            deleted_errors += n\n",
    "            continue\n",
    "            \n",
    "        del_positions = deletions2[(read, read_type)] if (read, read_type) in deletions2 else {}\n",
    "        err_positions = errors2[(read, read_type)] if (read, read_type) in errors2 else {}\n",
    "        for pos in positions.keys():\n",
    "            if pos in del_positions:\n",
    "                deleted_errors += 1\n",
    "            elif pos not in err_positions:\n",
    "                corrected_errors += 1\n",
    "            \n",
    "    new_errors = 0\n",
    "    for (read, read_type), positions in errors2.items():\n",
    "        err_positions = errors1[(read, read_type)] if (read, read_type) in errors1 else {}\n",
    "        for pos in positions.keys():\n",
    "            if pos not in err_positions:\n",
    "                new_errors += 1\n",
    "    deleted_good_n = deleted_correct_positions(errors1, deletions2, second_deleted_reads, sam1)\n",
    "    \n",
    "    result = pd.DataFrame(columns=['Error in corrected reads', 'Correct base in corrected reads', \n",
    "                                   'Base is absent in corrected reads'], \n",
    "                          index=['Error in raw data', 'Correct base in raw data'])\n",
    "    \n",
    "    result.loc['Error in raw data', 'Error in corrected reads'] = f'{n_errors - deleted_errors - corrected_errors} ({(n_errors - deleted_errors - corrected_errors) / n_errors * 100:.3f}%)'\n",
    "    \n",
    "    result.loc['Error in raw data', 'Correct base in corrected reads'] = f'{corrected_errors} ({corrected_errors / n_errors * 100:.3f}%)'\n",
    "    \n",
    "    result.loc['Error in raw data', 'Base is absent in corrected reads'] = f'{deleted_errors} ({deleted_errors / n_errors * 100:.3f}%)'\n",
    "    \n",
    "    result.loc['Correct base in raw data', 'Error in corrected reads'] = f'{new_errors} ({new_errors / good1 * 100:.3f}%)'\n",
    "    \n",
    "    result.loc['Correct base in raw data', 'Correct base in corrected reads'] = f'{good1 - deleted_good_n - new_errors} ({(good1 - deleted_good_n - new_errors) / good1 * 100:.3f}%)'\n",
    "    \n",
    "    result.loc['Correct base in raw data', 'Base is absent in corrected reads'] = f'{deleted_good_n} ({deleted_good_n / good1 * 100:.3f}%)'\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_start(sam):\n",
    "    read2position = {}\n",
    "    for num, line in enumerate(open(sam)):\n",
    "        if line.startswith('@'):\n",
    "            continue\n",
    "            \n",
    "        data = line.strip().split()\n",
    "        if data[2] == '*' or data[5] == '*':\n",
    "            continue\n",
    "            \n",
    "        start = int(data[3])\n",
    "        if not start:\n",
    "            continue\n",
    "            \n",
    "        read_name = data[0]\n",
    "        read_type = 'forward' if int(data[8]) > 0 else 'backward'\n",
    "        read2position[(read_name, read_type)] = start\n",
    "    return read2position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "== Warning ==  No assembly mode was sepcified! If you intend to assemble high-coverage multi-cell/isolate data, use '--isolate' option.\n",
      "\n",
      "\n",
      "Command line: /home/alex/anaconda3/bin/spades.py\t-1\t/home/alex/PycharmProjects/NGS/homework_4/data/ecoli_400K_err_1.fastq\t-2\t/home/alex/PycharmProjects/NGS/homework_4/data/ecoli_400K_err_2.fastq\t-o\t/home/alex/PycharmProjects/NGS/homework_4/spades_ecoli\t--only-error-correction\t--threads\t6\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 3.14.0\n",
      "  Python version: 3.7.4\n",
      "  OS: Linux-5.4.0-33-generic-x86_64-with-debian-bullseye-sid\n",
      "\n",
      "Output dir: /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli\n",
      "Mode: ONLY read error correction (without assembling)\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Standard mode\n",
      "  For multi-cell/isolate data we recommend to use '--isolate' option; for single-cell MDA data use '--sc'; for metagenomic data use '--meta'; for RNA-Seq use '--rna'.\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/home/alex/PycharmProjects/NGS/homework_4/data/ecoli_400K_err_1.fastq']\n",
      "      right reads: ['/home/alex/PycharmProjects/NGS/homework_4/data/ecoli_400K_err_2.fastq']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Read error correction parameters:\n",
      "  Iterations: 1\n",
      "  PHRED offset will be auto-detected\n",
      "  Corrected reads will be compressed\n",
      "Other parameters:\n",
      "  Dir for temp files: /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/tmp\n",
      "  Threads: 6\n",
      "  Memory limit (in Gb): 10\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/spades.log\n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "== Running: /home/alex/anaconda3/bin/spades-hammer /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/corrected/configs/config.info\n",
      "\n",
      "  0:00:00.000     4M / 11M   INFO    General                 (main.cpp                  :  75)   Starting BayesHammer, built from N/A, git revision N/A\n",
      "  0:00:00.000     4M / 11M   INFO    General                 (main.cpp                  :  76)   Loading config from /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/corrected/configs/config.info\n",
      "  0:00:00.002     4M / 11M   INFO    General                 (main.cpp                  :  78)   Maximum # of threads to use (adjusted due to OMP capabilities): 6\n",
      "  0:00:00.002     4M / 11M   INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 10 Gb\n",
      "  0:00:00.002     4M / 11M   INFO    General                 (main.cpp                  :  86)   Trying to determine PHRED offset\n",
      "  0:00:00.002     4M / 11M   INFO    General                 (main.cpp                  :  92)   Determined value is 33\n",
      "  0:00:00.002     4M / 11M   INFO    General                 (hammer_tools.cpp          :  38)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]\n",
      "  0:00:00.002     4M / 11M   INFO    General                 (main.cpp                  : 113)   Size of aux. kmer data 24 bytes\n",
      "     === ITERATION 0 begins ===\n",
      "  0:00:00.003     4M / 11M   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:00.003     4M / 11M   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 96 files using 6 threads. This might take a while.\n",
      "  0:00:00.003     4M / 11M   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:00.003     4M / 11M   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.555522 Gb\n",
      "  0:00:00.003     4M / 11M   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 699050\n",
      "  0:00:00.005     4G / 4G    INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /home/alex/PycharmProjects/NGS/homework_4/data/ecoli_400K_err_1.fastq\n",
      "  0:00:10.460     4G / 4G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 1381602 reads\n",
      "  0:00:10.460     4G / 4G    INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /home/alex/PycharmProjects/NGS/homework_4/data/ecoli_400K_err_2.fastq\n",
      "  0:00:17.736     4G / 4G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 2763204 reads\n",
      "  0:00:17.736     4G / 4G    INFO   K-mer Splitting          (kmer_data.cpp             : 112)   Total 2763204 reads processed\n",
      "  0:00:17.886   109M / 2G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:18.211   109M / 2G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 13488952 kmers in total.\n",
      "  0:00:18.211   109M / 2G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:18.307   109M / 2G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:19.164   115M / 2G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:00:19.661   115M / 2G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 6263040 bytes occupied (3.71447 bits per kmer).\n",
      "  0:00:19.695   115M / 2G    INFO   K-mer Counting           (kmer_data.cpp             : 351)   Arranging kmers in hash map order\n",
      "  0:00:20.428   333M / 2G    INFO    General                 (main.cpp                  : 148)   Clustering Hamming graph.\n",
      "  0:00:49.913   333M / 2G    INFO    General                 (main.cpp                  : 155)   Extracting clusters\n",
      "  0:00:55.909   320M / 2G    INFO    General                 (main.cpp                  : 167)   Clustering done. Total clusters: 2187352\n",
      "  0:00:55.916   208M / 2G    INFO   K-mer Counting           (kmer_data.cpp             : 371)   Collecting K-mer information, this takes a while.\n",
      "  0:00:56.045   528M / 2G    INFO   K-mer Counting           (kmer_data.cpp             : 377)   Processing /home/alex/PycharmProjects/NGS/homework_4/data/ecoli_400K_err_1.fastq\n",
      "  0:01:27.191   462M / 2G    INFO   K-mer Counting           (kmer_data.cpp             : 377)   Processing /home/alex/PycharmProjects/NGS/homework_4/data/ecoli_400K_err_2.fastq\n",
      "  0:01:58.459   462M / 2G    INFO   K-mer Counting           (kmer_data.cpp             : 384)   Collection done, postprocessing.\n",
      "  0:01:58.517   462M / 2G    INFO   K-mer Counting           (kmer_data.cpp             : 398)   There are 13488952 kmers in total. Among them 9831116 (72.8827%) are singletons.\n",
      "  0:01:58.517   462M / 2G    INFO    General                 (main.cpp                  : 173)   Subclustering Hamming graph\n",
      "  0:02:02.492   464M / 2G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 649)   Subclustering done. Total 1 non-read kmers were generated.\n",
      "  0:02:02.493   464M / 2G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 650)   Subclustering statistics:\n",
      "  0:02:02.493   464M / 2G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 651)     Total singleton hamming clusters: 1393328. Among them 2580 (0.185168%) are good\n",
      "  0:02:02.493   464M / 2G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 652)     Total singleton subclusters: 4146. Among them 4142 (99.9035%) are good\n",
      "  0:02:02.493   464M / 2G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 653)     Total non-singleton subcluster centers: 855746. Among them 836834 (97.79%) are good\n",
      "  0:02:02.493   464M / 2G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 654)     Average size of non-trivial subcluster: 14.1346 kmers\n",
      "  0:02:02.493   464M / 2G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 655)     Average number of sub-clusters per non-singleton cluster: 1.08295\n",
      "  0:02:02.493   464M / 2G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 656)     Total solid k-mers: 843556\n",
      "  0:02:02.494   464M / 2G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 657)     Substitution probabilities: [4,4]((0.925366,0.0417851,0.0246901,0.00815845),(0.0168258,0.960386,0.0122683,0.0105199),(0.0104921,0.0122447,0.960405,0.016858),(0.00814236,0.0246651,0.0417487,0.925444))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:02:02.497   464M / 2G    INFO    General                 (main.cpp                  : 178)   Finished clustering.\n",
      "  0:02:02.497   464M / 2G    INFO    General                 (main.cpp                  : 197)   Starting solid k-mers expansion in 6 threads.\n",
      "  0:02:23.677   461M / 2G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 0 produced 45116 new k-mers.\n",
      "  0:02:44.864   457M / 2G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 1 produced 1684 new k-mers.\n",
      "  0:03:06.171   457M / 2G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 2 produced 63 new k-mers.\n",
      "  0:03:32.174   457M / 2G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 3 produced 32 new k-mers.\n",
      "  0:04:03.383   457M / 2G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 4 produced 0 new k-mers.\n",
      "  0:04:03.383   457M / 2G    INFO    General                 (main.cpp                  : 222)   Solid k-mers finalized\n",
      "  0:04:03.384   457M / 2G    INFO    General                 (hammer_tools.cpp          : 222)   Starting read correction in 6 threads.\n",
      "  0:04:03.384   457M / 2G    INFO    General                 (hammer_tools.cpp          : 235)   Correcting pair of reads: /home/alex/PycharmProjects/NGS/homework_4/data/ecoli_400K_err_1.fastq and /home/alex/PycharmProjects/NGS/homework_4/data/ecoli_400K_err_2.fastq\n",
      "  0:04:06.078   910M / 2G    INFO    General                 (hammer_tools.cpp          : 170)   Prepared batch 0 of 600000 reads.\n",
      "  0:04:18.888   930M / 2G    INFO    General                 (hammer_tools.cpp          : 177)   Processed batch 0\n",
      "  0:04:20.832   930M / 2G    INFO    General                 (hammer_tools.cpp          : 187)   Written batch 0\n",
      "  0:04:23.032   930M / 2G    INFO    General                 (hammer_tools.cpp          : 170)   Prepared batch 1 of 600000 reads.\n",
      "  0:04:35.758   943M / 2G    INFO    General                 (hammer_tools.cpp          : 177)   Processed batch 1\n",
      "  0:04:37.490   943M / 2G    INFO    General                 (hammer_tools.cpp          : 187)   Written batch 1\n",
      "  0:04:38.116   943M / 2G    INFO    General                 (hammer_tools.cpp          : 170)   Prepared batch 2 of 181602 reads.\n",
      "  0:04:42.920   946M / 2G    INFO    General                 (hammer_tools.cpp          : 177)   Processed batch 2\n",
      "  0:04:43.721   946M / 2G    INFO    General                 (hammer_tools.cpp          : 187)   Written batch 2\n",
      "  0:04:43.914   786M / 2G    INFO    General                 (hammer_tools.cpp          : 276)   Correction done. Changed 883912 bases in 651938 reads.\n",
      "  0:04:43.915   786M / 2G    INFO    General                 (hammer_tools.cpp          : 277)   Failed to correct 21928 bases out of 252262696.\n",
      "  0:04:43.957   354M / 2G    INFO    General                 (main.cpp                  : 255)   Saving corrected dataset description to /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/corrected/corrected.yaml\n",
      "  0:04:43.962   354M / 2G    INFO    General                 (main.cpp                  : 262)   All done. Exiting.\n",
      "\n",
      "===== Read error correction finished. \n",
      "\n",
      "\n",
      "===== corrected reads compression started. \n",
      "\n",
      "\n",
      "== Running: /home/alex/anaconda3/bin/python /home/alex/anaconda3/share/spades/spades_pipeline/scripts/compress_all.py --input_file /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/corrected/corrected.yaml --ext_python_modules_home /home/alex/anaconda3/share/spades --max_threads 6 --output_dir /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/corrected --gzip_output\n",
      "\n",
      "== Compressing corrected reads (with gzip)\n",
      "\n",
      "===== corrected reads compression finished. \n",
      "\n",
      "\n",
      "===== Read error correction finished. \n",
      "\n",
      "\n",
      "===== Breaking scaffolds started. \n",
      "\n",
      "\n",
      "== Running: /home/alex/anaconda3/bin/python /home/alex/anaconda3/share/spades/spades_pipeline/scripts/breaking_scaffolds_script.py --result_scaffolds_filename /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/scaffolds.fasta --misc_dir /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/misc --threshold_for_breaking_scaffolds 3\n",
      "\n",
      "\n",
      "===== Breaking scaffolds finished. \n",
      "\n",
      "\n",
      "===== Terminate started. \n",
      "\n",
      "\n",
      "===== Terminate finished. \n",
      "\n",
      " * Corrected reads are in /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/corrected/\n",
      "\n",
      "======= SPAdes pipeline finished.\n",
      "\n",
      "SPAdes log can be found here: /home/alex/PycharmProjects/NGS/homework_4/spades_ecoli/spades.log\n",
      "\n",
      "Thank you for using SPAdes!\n"
     ]
    }
   ],
   "source": [
    "!spades.py -1 data/ecoli_400K_err_1.fastq -2 data/ecoli_400K_err_2.fastq -o spades_ecoli --only-error-correction \\\n",
    "--threads 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48352, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.27, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.047 CPU sec, 2.854 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48401, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.22, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.239 CPU sec, 2.898 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48399, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.944 CPU sec, 2.617 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48377, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.38, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.834 CPU sec, 2.544 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48371, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.847 CPU sec, 2.547 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48414, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.852 CPU sec, 2.552 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48317, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.850 CPU sec, 2.548 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48419, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.34, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.828 CPU sec, 2.537 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48404, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.40, 10.07)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.821 CPU sec, 2.527 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48388, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.36, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.840 CPU sec, 2.537 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48318, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.18)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.857 CPU sec, 2.544 real sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48338, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.30, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.846 CPU sec, 2.540 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48261, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.830 CPU sec, 2.550 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48332, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.35, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.841 CPU sec, 2.551 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48379, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.839 CPU sec, 2.553 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48396, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.29, 10.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.834 CPU sec, 2.569 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48371, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.45, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.830 CPU sec, 2.532 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48421, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.38, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.794 CPU sec, 2.490 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48381, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (179, 249)\n",
      "[M::mem_pestat] mean and std.dev: (214.26, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (165, 263)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.815 CPU sec, 2.507 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48426, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.181 CPU sec, 2.876 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48336, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.311 CPU sec, 2.893 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48423, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.24, 10.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.777 CPU sec, 2.502 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48403, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.32, 10.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mem_process_seqs] Processed 100000 reads in 2.761 CPU sec, 2.485 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48382, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 3.621 CPU sec, 3.291 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48355, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.31, 10.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.884 CPU sec, 2.599 real sec\n",
      "[M::process] read 100000 sequences (10000000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48327, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.25, 10.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.897 CPU sec, 2.608 real sec\n",
      "[M::process] read 16054 sequences (1605400 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 48388, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.28, 10.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 100000 reads in 2.665 CPU sec, 2.533 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 7766, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (207, 214, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (177, 252)\n",
      "[M::mem_pestat] mean and std.dev: (214.28, 10.28)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (162, 267)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 16054 reads in 0.553 CPU sec, 0.451 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem data/MG1655-K12.first400K.fasta spades_ecoli/corrected/ecoli_400K_err_1.00.0_0.cor.fastq.gz spades_ecoli/corrected/ecoli_400K_err_2.00.0_0.cor.fastq.gz\n",
      "[main] Real time: 72.467 sec; CPU: 79.656 sec\n"
     ]
    }
   ],
   "source": [
    "!bwa mem data/MG1655-K12.first400K.fasta spades_ecoli/corrected/ecoli_400K_err_1.00.0_0.cor.fastq.gz \\\n",
    "spades_ecoli/corrected/ecoli_400K_err_2.00.0_0.cor.fastq.gz > spades_ecoli.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "start_positions = read_start('ecoli_400.sam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "spades_errors, spades_deletions, spades_correct_positions = cnt_err_del('spades_ecoli.sam',\n",
    "                                                                        'data/MG1655-K12.first400K.fasta',\n",
    "                                                                        start_positions=start_positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = spades_table(raw_errors, spades_errors, spades_deletions, 'ecoli_400.sam', 'spades_ecoli.sam', \n",
    "                      raw_correct_positions, spades_correct_positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Error in corrected reads</th>\n",
       "      <th>Correct base in corrected reads</th>\n",
       "      <th>Base is absent in corrected reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Error in raw data</td>\n",
       "      <td>7605795 (51.213%)</td>\n",
       "      <td>826684 (5.566%)</td>\n",
       "      <td>6418941 (43.221%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Correct base in raw data</td>\n",
       "      <td>26086 (0.010%)</td>\n",
       "      <td>241818403 (92.591%)</td>\n",
       "      <td>19324458 (7.399%)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Error in corrected reads  \\\n",
       "Error in raw data               7605795 (51.213%)   \n",
       "Correct base in raw data           26086 (0.010%)   \n",
       "\n",
       "                         Correct base in corrected reads  \\\n",
       "Error in raw data                        826684 (5.566%)   \n",
       "Correct base in raw data             241818403 (92.591%)   \n",
       "\n",
       "                         Base is absent in corrected reads  \n",
       "Error in raw data                        6418941 (43.221%)  \n",
       "Correct base in raw data                 19324458 (7.399%)  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# вывод: кажется, что Trimmamatic справляется лучше, так как TP значительно выше, чем у Spades, TN мерно \n",
    "# одинаковы."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
